rfcs/20211118-support-implicit-concat/README.md

```diff
@@ -0,0 +1,132 @@
+
+1. Setup primitive descriptors of primitives which would contribute to concat.
+   a. If any of those are compute-intense primitive, querying of destination
+      memory descriptor would be required to get a proper memory format.
+   b. If all of those are memory-bound primitives, query is not required,
+      formats should already be defined and coincide or be compatible.
+2. Setup concat primitive descriptor with destination descriptors from previous
+   step. Omit concat destination to let the library deduce it.
+   a. If everything is good to go, destination would have same blocked format.
+   b. If not, concat would return `abcd`-like format.
+3. Verify that destination was created properly. This can be achieved by
+   creating a memory descriptor with concat destination dims and plain
+   `abcd`-like layout. If they match with queried, concat failed to prepare an
+   implicit setup. Consider using an explicit one.
+4. Create submemory descriptors from concat destination memory descriptor.
+   a. If both memory descriptors were created properly, it means that implicit
+      concat is good to go.
+   b. It may happen that not best optimized implementation is taken, but such
+      situations should be treated as the library problems and reported.
+5. Re-create primitive descriptors of contributing primitives.
+   a. Ideally, implementations with dense and submemoried destination
+      descriptors should coincide. They may be additionally validated with
+      `impl_info_str()` call.
+6. Create new memory objects for implicit concat destination and for
+   contributing primitives. Last one should utilize a handle from the first one.
+   Pass updated memory objects for execution.
+
+This algorithm is demonstrated at [this example](implicit_concat.cpp). Example
+has coverage for both scenarios depending on values provided for IC, OC, and N.
+
+### Coverage
+
+Since this feature is complex enough to maintain, support and validate,
+primitive support will be extended upon request. The initial implementation
+would support convolution and pooling with strided input and output.
+
+## Testing
+
+Testing is likely to be extended with an additional flag which would define a
+bigger tensor over specified dimension(s). Then, primitive descriptor would be
+created as the leftmost or rightmost part of it (validating offset0 application
+correctness).
+
+Preliminary interface is like this:
+```bash
+benchdnn --conv --submemory=src:2x128x3x3:0x0x0x0+dst:2x256x3x3:0x192x0x0
+                mb2_ic32oc64_ih3oh3kh3ph1sh1
+```
+
+At the example above, interface is breaking down in three components:
+* argument to apply submemory descriptor.
+* greater memory dimensions to create submemory for a problem.
+* offsets.
+
+Each of argument will be able to get a submemory separated by `+`. We expect
+users to focus on case where channels would be a primary axis of submemory but
+interface is not limited to this case. Though, implementations may.
```

Comment:
"...to this case. Though,..." --> "...to this case, though some implementations may."

Reply:
Thanks for a suggestion, applied.